RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGeneratorGPU.cc

```diff
@@ -0,0 +1,460 @@
+      v.clear();
+  }
+}
+void fillGraph(const SeedingLayerSetsHits &layers,
+               const IntermediateHitDoublets::RegionLayerSets &regionLayerPairs,
+               CAGraph &g, std::vector<const HitDoublets *> &hitDoublets) {
+
+  for (unsigned int i = 0; i < layers.size(); i++) {
+    for (unsigned int j = 0; j < 4; ++j) {
+      auto vertexIndex = 0;
+      auto foundVertex = std::find(g.theLayers.begin(), g.theLayers.end(),
+                                   layers[i][j].name());
+      if (foundVertex == g.theLayers.end()) {
+        vertexIndex = g.theLayers.size() - 1;
+      } else {
+        vertexIndex = foundVertex - g.theLayers.begin();
+      }
+
+      if (j > 0) {
+
+        auto innerVertex = std::find(g.theLayers.begin(), g.theLayers.end(),
+                                     layers[i][j - 1].name());
+
+        CALayerPair tmpInnerLayerPair(innerVertex - g.theLayers.begin(),
+                                      vertexIndex);
+
+        if (std::find(g.theLayerPairs.begin(), g.theLayerPairs.end(),
+                      tmpInnerLayerPair) == g.theLayerPairs.end()) {
+          auto found = std::find_if(
+              regionLayerPairs.begin(), regionLayerPairs.end(),
+              [&](const IntermediateHitDoublets::LayerPairHitDoublets &pair) {
+                return pair.innerLayerIndex() == layers[i][j - 1].index() &&
+                       pair.outerLayerIndex() == layers[i][j].index();
+              });
+          if (found != regionLayerPairs.end()) {
+            hitDoublets.emplace_back(&(found->doublets()));
+            g.theLayerPairs.push_back(tmpInnerLayerPair);
+            g.theLayers[vertexIndex].theInnerLayers.push_back(
+                innerVertex - g.theLayers.begin());
+            innerVertex->theOuterLayers.push_back(vertexIndex);
+            g.theLayers[vertexIndex].theInnerLayerPairs.push_back(
+                g.theLayerPairs.size() - 1);
+            innerVertex->theOuterLayerPairs.push_back(g.theLayerPairs.size() -
+                                                      1);
+          }
+        }
+      }
+    }
+  }
+
+  //   std::cout << "number of layer pairs " << hitDoublets.size() << std::endl;
+  //
+  // for (unsigned int i = 0; i < hitDoublets.size(); ++i) {
+  //   std::cout << i << " " << hitDoublets[i]->size() << std::endl;
+  // }
+  //
+  // for (unsigned int i = 0; i < g.theLayerPairs.size(); ++i) {
+  //   std::cout << "layer pair " << i << " "
+  //             << g.theLayers[g.theLayerPairs[i].theLayers[0]].name()
+  //             << " hits: " << hitDoublets[i]->innerLayer().size() << " and "
+  //             << g.theLayers[g.theLayerPairs[i].theLayers[1]].name()
+  //             << " hits: " << hitDoublets[i]->outerLayer().size() <<
+  //             std::endl;
+  // }
+}
+} // namespace
+
+void CAHitQuadrupletGeneratorGPU::hitNtuplets(
+    const IntermediateHitDoublets &regionDoublets,
+    std::vector<OrderedHitSeeds> &result, const edm::EventSetup &es,
```

Comment:
Why do you need to pass `result` here? I don't see it used. Removing would allow removing the `ntuplets` member of `CAHitNtupletHeterogeneousEDProducer`.

Reply:
@felicepantaleo can you comment on this point ?